jina/drivers/rank/__init__.py

```diff
@@ -82,29 +76,36 @@ def _apply_all(self, docs: 'DocumentSet', *args, **kwargs) -> None:
             )
 
             matches = doc.matches
-            old_match_scores = {match.id: match.score.value for match in matches}
-            match_meta = (
-                {match.id: match.get_attrs(*self._exec_match_keys) for match in matches}
-                if self._exec_match_keys
-                else None
-            )
+            num_matches = len(matches)
+            old_match_scores = []
+            needs_match_meta = self._exec_match_keys is not None
+            match_meta = [] if needs_match_meta else None
+            for match in matches:
+                old_match_scores.append(match.score.value)
+                if needs_match_meta:
+                    match_meta.append(match.get_attrs(*self._exec_match_keys))
 
             # if there are no matches, no need to sort them
             if not old_match_scores:
                 continue
 
-            new_match_scores = self.exec_fn(query_meta, old_match_scores, match_meta)
-            self._sort_matches_in_place(doc, new_match_scores)
+            new_scores = self.exec_fn(old_match_scores, query_meta, match_meta)
+            if num_matches != len(new_scores):
+                msg = (
+                    f'The number of matches to be scored {num_matches} do not match the number of scores returned '
+                    f'by the ranker {self.exec.__name__} '
+                )
+                self.logger.error(msg)
+                raise ValueError(msg)
+            self._sort_matches_in_place(matches, new_scores)
 
     def _sort_matches_in_place(
-        self, context_doc: 'Document', match_scores: 'np.ndarray'
+        self, matches: 'MatchSet', match_scores: Iterable[float]
     ) -> None:
         op_name = self.exec.__class__.__name__
-        cm = context_doc.matches
-        cm.build()
-        for match_id, score in match_scores:
-            cm[match_id].score = NamedScore(
-                value=score, op_name=op_name, ref_id=context_doc.id
-            )
+        ref_doc_id = matches._ref_doc.id
+
+        for match, score in zip(matches, match_scores):
+            match.score = NamedScore(value=score, op_name=op_name, ref_id=ref_doc_id)
 
-        cm.sort(key=lambda x: x.score.value, reverse=True)
+        matches.sort(key=lambda x: x.score.value, reverse=True)
```

Comment:
```suggestion
        matches.sort(key=lambda x: x.score.value, reverse=self.reverse)
```

Reply:
We do not have a `reverse` parameter. I think  for now every `Ranker` is scoring per `relevance` so it is fair to set it to `True`

Reply:
We do have. 
https://github.com/jina-ai/jina/pull/2113/files/bade8af72d03b6690bdee454907f49c676283125#diff-5263f3d40f3d59bd6ea5d52f39b61f12a708f1e44ec18487347bfc57b803bea0R60

Reply:
```
ds = DocumentSet([Document(text=f'{np.random.randint(0, 10)}') for i in range(3)])
ds.sort(key=lambda x: dunder_get(x, 'text'), reverse=True)
# {'id': '1040e3b6-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '9'}
# {'id': '1040ec8a-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '7'}
# {'id': '1040e8d4-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '4'}

ds.sort(key=lambda x: x.text, reverse=False)
# {'id': '1040e8d4-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '4'}
# {'id': '1040ec8a-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '7'}
# {'id': '1040e3b6-808e-11eb-a397-00e04c6b0562', 'mimeType': 'text/plain', 'text': '9'}
```